### ISSUE TITLE
Available connections is suddenly 0 without recovering in 0.12

### ISSUE THREAD

**Issue by dagingaa** [tgriesser/knex#1688]
Hi all!

We're seeing a weird issue recently, where available connections suddenly drop to 0 without being able to recover on one of our servers. Other servers are chugging along nicely, but every once in a while this happens. The consequence is that database queues grow indefinitely and the usual `Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?` is thrown all over the place. I've checked our code for nasty transactions as suggested, but we don't have any that could explain suddenly locking up 10 database connections.

*In New Relic we see the following pattern for available connections*:

![image](https://cloud.githubusercontent.com/assets/284996/18786905/985c4e08-81a0-11e6-9962-bf86440921e6.png)

As you can see, one second we have 10 available, and the next 0. This does not get fixed until I restart the process, which is what you see when it goes back up to 10. This happens suddenly and seemingly without any obvious causes. However, we are handling a lot of traffic on these machines, so the log can be quite chaotic, especially since we're most likely looking for that one code path that went wrong, needle in a haystack, if you will.

To me, this looks like Knex suddenly mass disconnects, but we only see this problem on one server fleet, and not on the other fleet. These run different code, which suggest the error lies with our code and not Knex, but at this point I'm desperate to find a solution. I have read through thousand of lines of code looking for a culprit, to no avail. We can bandaid the problem by marking the server as unhealthy, but I would much rather understand why this happens, and how we can avoid it in the future.

Has anyone seen anything like this before? Any hunches to where I should continue looking?

To help debug this issue, I'm attaching some code examples of how we do transactions today, perhaps we're doing something horribly wrong there and I'm just not picking up on it.

```
.transaction((t) => {
    return Promise.all([
        t.delete().from("table1").where("field", fieldValue),
        t.delete().from("table2").where("field", fieldValue)
    ]);
})
```

Thank you to whoever took the time to read through it, all help is appreciated!

**Comment by tgriesser** [comment]
What database are you using?

**Comment by tgriesser** [comment]
Also is there a chance you'd be calling `knex.destroy()` anywhere in your codebase (perhaps inadvertently)

**Comment by dagingaa** [comment]
We're using postgresql. Forgot to mention that!

**Comment by tgriesser** [comment]
@dagingaa also if you could run with `DEBUG=knex:pool` you might get more information about what's going on... or (undocumented and likely to change but works on 0.12.x) if you pass a "log" function as the pool config:

```js
const knex = require('knex')
const db = knex({
  client: 'pg',
  connection: ...,
  pool: {
    min: 0,
    max: 10,
    log(msg, level) {
      // handle pool logging here
    }
  }
})

```

It'll catch more of the logs in [generic-pool](https://github.com/coopernurse/node-pool/blob/b89166bd10eb3e28f7499c394cde8cace8b9fbb6/lib/generic-pool.js#L165). Currently the `DEBUG=knex:pool` only captures the "info" level of logging [here](https://github.com/tgriesser/knex/blob/67b62d3213aa6ade68bdf84cb95b3410a1fd8053/src/client.js#L171), not the "verbose" log which might narrow things down a bit.

**Comment by tgriesser** [comment]
But that is strange, other than calling `knex.destroy` prematurely, I'm not sure what would cause all of the connections to drop like that. Is this only happening in 0.12 and not other versions?

**Comment by dagingaa** [comment]
We've seen (we think, we've lost the logs to be able to debug that now) this issue happening in 0.11 as well. However, since I didn't think to check the queues at that time, I don't have concrete data on it. We did see the Knex aquireConnectionTimeout error fire as noted above though. We had that issue last Friday and throughout the weekend before we made the switch to 0.12 where stuff has been fine until about late last night and today when this is happening suddenly and with no warning. This was based on us finding an issue I've lost the link to which noted that pool2 had some issues, and that generic-pool was being used in 0.12. That assumptions seems to have been wrong (but it's so good to be able to blame someone else for a change, and not our own bad choices).

I checked our code path for `destroy()` but didn't find anything that calls it except some integration tests. 

I'll try running with extended logging and see what we get from that, good idea! I'll come back to you once I have some more data to look at.

**Comment by tgriesser** [comment]
Nice 👍

**Comment by dagingaa** [comment]
I identified the issue on our end. As suspected, this had nothing to do with knex but our code. We had some ugly queries that for reasons unknown to me got "stuck" in postgresql for several minutes. If a server got hit with multiple requests, which didn't happen that often, the connection pool would suddenly drain of available connections, meaning no other queries could be run. 

I debugged the issues by seeing the issue in production, then going to the database manually and running:

`SELECT query, client_addr, query_start FROM pg_stat_activity ORDER BY query_start LIMIT 50;`

and found 10 identical queries, all from the same server, having query_start times more than 8 minutes ago.

I'm unsure why these queries weren't killed on some kind of timeout to avoid this issue, but then again, that would have masked the issue for us for even longer. I'm also not sure at the moment why the query sometimes worked, and sometimes, seemingly at once decided to lock up. I did not find any active deadlocks in place, but then again, I'm no expert in debugging database issues.

I hope the following issue may help others in their debugging in the future. Special thanks to @tgriesser for responding so quickly and giving me some ideas on how to debug this issue. We really appreciate your work! Love from the appear.in team :)
